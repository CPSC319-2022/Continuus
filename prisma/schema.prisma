// This is your Prisma schema file,
// learn more about it in the docs: https://pris.ly/d/prisma-schema

generator client {
    provider = "prisma-client-js"
}

generator zod {
    provider = "prisma-zod-generator"
    output = "../src/generated"
}

datasource db {
    provider = "postgres"
    url      = env("DATABASE_URL")
}

model Example {
    id        String   @id @default(cuid())
    createdAt DateTime @default(now())
    updatedAt DateTime @updatedAt
}

// Necessary for Next auth
model Account {
    id                String  @id @default(cuid())
    userId            String
    type              String
    provider          String
    providerAccountId String
    refresh_token     String?
    access_token      String?
    expires_at        Int?
    token_type        String?
    scope             String?
    id_token          String?
    session_state     String?
    user              User    @relation(fields: [userId], references: [id], onDelete: Cascade)

    @@unique([provider, providerAccountId])
}

model Session {
    id           String   @id @default(cuid())
    sessionToken String   @unique
    userId       String
    expires      DateTime
    user         User     @relation(fields: [userId], references: [id], onDelete: Cascade)
}

model User {
    id            String     @id @default(cuid())
    name          String?
    email         String?    @unique
    emailVerified DateTime?
    image         String?
    role          Role       @default(READER)
    accounts      Account[]
    sessions      Session[]
    BlogPost      BlogPost[]
}

model VerificationToken {
    identifier String
    token      String   @unique
    expires    DateTime

    @@unique([identifier, token])
}

model BlogPost {
    id          String @id @default(cuid())
    userId      String
    title       String
    content     String
    user        User   @relation(fields: [userId], references: [id], onDelete: Cascade)
    comments    Comment[]
}

model Comment {
    id      String @id @default(cuid())
    content String
    user    User   @relation(fields: [userId], references: [id], onDelete: Cascade)
}

enum Role {
    READER
    CONTRIBUTOR
    ADMIN
}
