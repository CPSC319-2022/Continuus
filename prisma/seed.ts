import { BlogPost, PrismaClient, User } from "@prisma/client";
import { Role } from "@prisma/client";
import { sentence, article } from "txtgen";
import { uniqueNamesGenerator, Config, names } from "unique-names-generator";

const config: Config = {
  dictionaries: [names],
};

const users: User[] = [];
const blogposts: BlogPost[] = [];

const prisma = new PrismaClient();

const main = async () => {
  console.log(`Start seeding ...`);

  // User
  console.log("Creating Users...");
  for (let i = 0; i < 20; i++) {
    const name: string = uniqueNamesGenerator(config);
    const role: Role = generateRole(Role);

    const user = await prisma.user.create({
      data: {
        name: name,
        email: `${name}@mailinator.com`,
        image: `https://i.pravatar.cc/150?img=${i}`,
        role: role,
      },
    });
    users.push(user);
  }
  console.log("20 users created");

  // get contributors and admins
  const contributorsAndAdmins: User[] = users.filter(({ role }) => {
    return role === Role.ADMIN || role === Role.CONTRIBUTOR;
  });

  const getIdOfRandomContributorOrAdmin = (): string => {
    const values = Object.values(contributorsAndAdmins);
    const randomIndex = Math.floor(Math.random() * values.length);
    return values[randomIndex]!.id;
  };

  // BlogPost
  console.log("Creating Blog Posts...");
  for (let _ = 0; _ < 100; _++) {
    const userId: string = getIdOfRandomContributorOrAdmin();
    const title: string = generateRandomTitle();
    const blogContent: string = generateRandomMarkdown();

    const blogpost = await prisma.blogPost.create({
      data: {
        userId: userId,
        title: title,
        content: blogContent,
      },
    });
    blogposts.push(blogpost);
  }
  console.log("100 blog posts created");

  // Comment
  console.log("Creating Comments...");
  blogposts.map(async ({ id }) => {
    const comments: string[] = generateComments(50);
    for (let i = 0; i < comments.length; i++) {
      await prisma.comment.create({
        data: {
          userId: getIdOfRandomContributorOrAdmin(),
          blogPostId: id,
          content: comments[i]!,
        },
      });
    }
  });
  console.log("0 to 50 comments created for each blog post");
};

const generateRole = (Role: {
  READER: "READER";
  CONTRIBUTOR: "CONTRIBUTOR";
  ADMIN: "ADMIN";
}) => {
  const values = Object.values(Role);
  const randomIndex = Math.floor(Math.random() * values.length);
  return values[randomIndex]!;
};

const generateRandomTitle = (): string => {
  const randomIndex = Math.floor(Math.random() * 3) + 3; // 3 to 5 inclusive
  const title = sentence().split(" ").slice(0, randomIndex).join(" ");
  if (title.endsWith(",")) {
    return title.slice(0, title.length - 1);
  }
  return title;
};

const generateRandomMarkdown = (): string => {
  const num = Math.floor(Math.random() * 2); // generates 0 or 1
  if (num == 1) {
    return `# ${sentence()}\n*${sentence()}*\n${article()}`;
  }
  return article();
};

// generates between 0 to maxNumber inclusive
const generateComments = (maxNumber: number): string[] => {
  // get random number between 0 and 50 inclusive
  const num = Math.floor(Math.random() * (maxNumber + 1));
  const comments: string[] = [];
  for (let i = 0; i < num; i++) {
    comments.push(sentence());
  }
  return comments;
};

main()
  .then(async () => {
    await prisma.$disconnect();
  })
  .catch(async (e) => {
    console.error(e);
    await prisma.$disconnect();
    process.exit(1);
  });
