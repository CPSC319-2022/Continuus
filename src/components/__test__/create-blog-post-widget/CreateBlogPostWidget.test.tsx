import { describe, it, expect, beforeEach, vi } from 'vitest'
import { cleanup, render, screen } from "@testing-library/react";
import userEvent from '@testing-library/user-event'
import { User } from '.prisma/client';

describe("Snapshot: CreateBlogPostWidget", () => {
    beforeEach(() => {
        cleanup()
        vi.resetModules();
    })

    const setup = (overrides?: {
        api: {
            blogPost?: {
                create: {
                    useMutation: (data: any) => ({
                        isSuccess: boolean,
                        isLoading: boolean,
                    })
                }
            },
            user?: {
                currentUser: {
                    useQuery: () => ({
                        data: User | null,
                        isLoading: boolean,
                    })
                }
            },
            useContext?: () => ({
                blogPost: {
                    get: {
                        invalidate: () => void
                    },
                    count: {
                        invalidate: () => void
                    }
                },
            })
        }
    }) => {
        vi.doMock("~/utils/api", () => ({
            api: {
                blogPost: {
                    create: {
                        useMutation: () => ({
                            isSuccess: false,
                            isLoading: false,
                        }),
                    }
                },
                user: {
                    currentUser: {
                        useQuery: () => ({
                            data: null,
                            isLoading: false,
                        }),
                    }
                },
                useContext: () => ({
                    blogPost: {
                        get: {
                            invalidate: vi.fn(),
                        },
                        count: {
                            invalidate: vi.fn()
                        }
                    },
                }),
                ...(overrides?.api)
            },
        }));
    }

    it("Unauthorized Plain CreateBlogPostWidget", async () => {
        setup();

        const { CreateBlogPostWidget } = await import('~/components/create-blog-post-widget');

        const body = render(<CreateBlogPostWidget />).baseElement;
        expect(body).toMatchSnapshot();
    });

    it("Loading CreateBlogPostWidget", async () => {
        setup({
            api: {
                user: {
                    currentUser: {
                        useQuery: () => ({
                            data: null,
                            isLoading: true
                        })
                    }
                }
            }
        })

        const { CreateBlogPostWidget } = await import('~/components/create-blog-post-widget');

        const body = render(<CreateBlogPostWidget />).baseElement;
        expect(body).toMatchSnapshot();
    });

    it("Authorized CreateBlogPostWidget", async () => {
        setup({
            api: {
                user: {
                    currentUser: {
                        useQuery: () => ({
                            data: { 
                                id: "1", 
                                email: "a", 
                                name: "hi mom", 
                                image: "c", 
                                emailVerified: null, 
                                role: "ADMIN", 
                                createdAt: new Date(), 
                                updatedAt: new Date()
                            },
                            isLoading: false
                        })
                    }
                }
            }
        })

        const { CreateBlogPostWidget } = await import('~/components/create-blog-post-widget');

        const body = render(<CreateBlogPostWidget />).baseElement;
        expect(body).toMatchSnapshot();
    });

    it("Authorized & Create Blog Post modal is open", async () => {
        setup({
            api: {
                user: {
                    currentUser: {
                        useQuery: () => ({
                            data: { 
                                id: "1", 
                                email: "a", 
                                name: "hi mom", 
                                image: "c", 
                                emailVerified: null, 
                                role: "ADMIN", 
                                createdAt: new Date(), 
                                updatedAt: new Date()
                            },
                            isLoading: false
                        })
                    }
                }
            }
        })

        const { CreateBlogPostWidget } = await import('~/components/create-blog-post-widget');

        const user = userEvent.setup()
        const body = render(<CreateBlogPostWidget />).baseElement;
        const actualButton = screen.getByRole("button");
        await user.click(actualButton)
        expect(body).toMatchSnapshot();
    });

    it("Authorized & Mutation is loading", async () => {
        setup({
            api: {
                user: {
                    currentUser: {
                        useQuery: () => ({
                            data: { 
                                id: "1", 
                                email: "a", 
                                name: "hi mom", 
                                image: "c", 
                                emailVerified: null, 
                                role: "ADMIN", 
                                createdAt: new Date(), 
                                updatedAt: new Date()
                            },
                            isLoading: false
                        })
                    }
                },
                blogPost: {
                    create: {
                        useMutation: () => ({
                            isSuccess: false,
                            isLoading: true,
                        }),
                    }
                },
            }
        })

        const { CreateBlogPostWidget } = await import('~/components/create-blog-post-widget');

        const user = userEvent.setup()
        const body = render(<CreateBlogPostWidget />).baseElement;
        const actualButton = screen.getByRole("button");
        await user.click(actualButton)
        expect(body).toMatchSnapshot();
    });

    it.skip("Authorized & Mutation is done", async () => {
      const blogPostCreateMutation = vi.fn(
        ({ onSuccess }: { onSuccess: () => void }) => {
          onSuccess();

          return {
            isSuccess: true,
            isLoading: false,
          };
        }
      );
      const invalidateBlogPostCount = vi.fn();
      const invalidateBlogPostGet = vi.fn();

      setup({
        api: {
          user: {
            currentUser: {
              useQuery: () => ({
                data: {
                  id: "1",
                  email: "a",
                  name: "hi mom",
                  image: "c",
                  emailVerified: null,
                  role: "ADMIN",
                  createdAt: new Date(),
                  updatedAt: new Date(),
                },
                isLoading: false,
              }),
            },
          },
          blogPost: {
            create: {
              useMutation: blogPostCreateMutation,
            },
          },
          useContext: () => ({
            blogPost: {
              count: {
                invalidate: invalidateBlogPostCount,
              },
              get: {
                invalidate: invalidateBlogPostGet,
              },
            },
          }),
        },
      });

      const { CreateBlogPostWidget } = await import(
        "~/components/create-blog-post-widget"
      );

      const user = userEvent.setup();
      const body = render(<CreateBlogPostWidget />).baseElement;
      const actualButton = screen.getByRole("button");
      await user.click(actualButton);
      expect(body).toMatchSnapshot();
      expect(invalidateBlogPostCount).toBeCalled();
      expect(invalidateBlogPostGet).toBeCalled();
    });
})