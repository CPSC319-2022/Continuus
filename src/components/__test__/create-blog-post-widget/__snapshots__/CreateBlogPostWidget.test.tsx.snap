// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Snapshot: CreateBlogPostWidget > Authorized & Create Blog Post modal is open 1`] = `
<body
  class="ReactModal__Body--open"
>
  <div
    data-react-modal-body-trap=""
    style="position: absolute; opacity: 0;"
    tabindex="0"
  />
  <div>
    <div
      aria-label="Create Blog Post Button"
      class="
                    fixed
                    bottom-24
                    left-1/2 
                    z-10
                    md:left-[95%] 
                    -translate-x-1/2
                    translate-y-1/2
                "
    >
      <div
        class="
                            absolute
                            w-64
                            min-h-0
                            h-12
                            -translate-y-full
                            -translate-x-1/2
                            left-1/2
                            md:-translate-x-full
                            md:left-full
                            overflow-hidden
                            -top-2
                        "
        style="height: 16rem;"
      >
        <div
          class="
                            top-full
                            absolute
                            w-full  
                            border-black
                            bg-gray-400
                            rounded-md
                            p-2
                            transition-all
                            after:absolute first-letter:
                            after:left-0
                            md:after:left-auto
                            after:right-0
                            after:my-0
                            after:mx-auto
                            after:w-0
                            after:h-0
                            after:border-t-[1rem]
                            after:border-solid
                            after:border-t-gray-400
                            after:border-l-[2rem]
                            after:border-r-[2rem]
                            after:border-x-transparent
                            after:transition-all
                        "
        >
          <p
            class="p-3 text-center md:text-left"
          >
            You do not have authorization to create posts
          </p>
        </div>
      </div>
      <button
        class="
                        bg-highlight-green hover:bg-gray-700 hover:scale-110 bg-highlight-green cursor-pointer transition-all 
                        aspect-square
                        w-16 
                        rounded-full 
                        after:block 
                        before:block 
                        after:bg-white 
                        before:bg-white
                        after:absolute 
                        before:absolute
                        after:top-1/2 
                        before:top-1/2 
                        after:left-1/2 
                        before:left-1/2 
                        after:-translate-x-1/2 
                        before:-translate-x-1/2 
                        after:-translate-y-1/2 
                        before:-translate-y-1/2 
                        after:h-8 
                        before:h-2 
                        after:w-2 
                        before:w-8 
                        after:transition-all 
                        before:transition-all 
                        "
      />
    </div>
  </div>
  <div
    class="ReactModalPortal"
  >
    <div
      class="ReactModal__Overlay fixed inset-0 z-20 bg-black/75"
    >
      <div
        aria-label="Blog Post Input"
        aria-modal="true"
        class="ReactModal__Content absolute top-1/2 left-1/2 z-40 w-11/12 -translate-x-1/2 -translate-y-1/2 rounded-md bg-white md:w-1/2 p-4"
        role="dialog"
        tabindex="-1"
      >
        <div
          class="flex pb-3 w-full flex-row items-center justify-between border-b border-b-gray-200 pl-4 pr-4"
        >
          <div
            class="font-semibold"
          >
            Create
             Blog Post
          </div>
          <button
            class="rounded-full transition-colors after:content-['\\\\2715'] hover:text-highlight-red hover:scale-125 transition-all"
          />
        </div>
        <form
          aria-label="Create Blog Post Form"
        >
          <div
            class="py-4 flex flex-row items-start"
          >
            <div
              class="flex flex-col w-full ml-2"
            >
              <input
                class="w-full rounded-md p-2 mb-4 text-2xl font-bold"
                placeholder="Write title here"
                type="text"
                value=""
              />
              <textarea
                class="w-full resize-none rounded-md h-96 p-2"
                placeholder="Write your blog post here"
              />
            </div>
          </div>
          <input
            class="relative left-full -translate-x-full h-10 rounded-md border w-32 text-center text-gray-700 disabled cursor-not-allowed bg-gray-200"
            type="submit"
            value="Create"
          />
        </form>
      </div>
    </div>
  </div>
  <div
    data-react-modal-body-trap=""
    style="position: absolute; opacity: 0;"
    tabindex="0"
  />
</body>
`;

exports[`Snapshot: CreateBlogPostWidget > Authorized & Mutation is done 1`] = `
<body
  class="ReactModal__Body--open"
>
  <div
    data-react-modal-body-trap=""
    style="position: absolute; opacity: 0;"
    tabindex="0"
  />
  <div
    class="ReactModalPortal"
  />
  <div
    class="ReactModalPortal"
  />
  <div>
    <div
      aria-label="Create Blog Post Button"
      class="
                    fixed
                    bottom-24
                    left-1/2 
                    z-10
                    md:left-[95%] 
                    -translate-x-1/2
                    translate-y-1/2
                "
    >
      <div
        class="
                            absolute
                            w-64
                            min-h-0
                            h-12
                            -translate-y-full
                            -translate-x-1/2
                            left-1/2
                            md:-translate-x-full
                            md:left-full
                            overflow-hidden
                            -top-2
                        "
        style="height: 16rem;"
      >
        <div
          class="
                            top-full
                            absolute
                            w-full  
                            border-black
                            bg-gray-400
                            rounded-md
                            p-2
                            transition-all
                            after:absolute first-letter:
                            after:left-0
                            md:after:left-auto
                            after:right-0
                            after:my-0
                            after:mx-auto
                            after:w-0
                            after:h-0
                            after:border-t-[1rem]
                            after:border-solid
                            after:border-t-gray-400
                            after:border-l-[2rem]
                            after:border-r-[2rem]
                            after:border-x-transparent
                            after:transition-all
                        "
        >
          <p
            class="p-3 text-center md:text-left"
          >
            You do not have authorization to create posts
          </p>
        </div>
      </div>
      <button
        class="
                        bg-highlight-green hover:bg-gray-700 hover:scale-110 bg-highlight-green cursor-pointer transition-all 
                        aspect-square
                        w-16 
                        rounded-full 
                        after:block 
                        before:block 
                        after:bg-white 
                        before:bg-white
                        after:absolute 
                        before:absolute
                        after:top-1/2 
                        before:top-1/2 
                        after:left-1/2 
                        before:left-1/2 
                        after:-translate-x-1/2 
                        before:-translate-x-1/2 
                        after:-translate-y-1/2 
                        before:-translate-y-1/2 
                        after:h-8 
                        before:h-2 
                        after:w-2 
                        before:w-8 
                        after:transition-all 
                        before:transition-all 
                        "
      />
    </div>
  </div>
  <div
    class="ReactModalPortal"
  >
    <div
      class="ReactModal__Overlay fixed inset-0 z-20 bg-black/75"
    >
      <div
        aria-label="Blog Post Input"
        aria-modal="true"
        class="ReactModal__Content absolute top-1/2 left-1/2 z-40 w-11/12 -translate-x-1/2 -translate-y-1/2 rounded-md bg-white md:w-1/2 p-4"
        role="dialog"
        tabindex="-1"
      >
        <div
          class="flex pb-3 w-full flex-row items-center justify-between border-b border-b-gray-200 pl-4 pr-4"
        >
          <div
            class="font-semibold"
          >
            Create
             Blog Post
          </div>
        </div>
        <div
          class="flex h-96 w-full flex-col items-center justify-center"
        >
          <svg
            class="animate-pulse stroke-highlight-green"
            fill="none"
            height="12rem"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="12rem"
            xmlns="http://www.w3.org/2000/svg"
          >
            <polyline
              points="20 6 9 17 4 12"
            />
          </svg>
          <div
            class="text-lg"
          >
            Successfully created the blog post
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    data-react-modal-body-trap=""
    style="position: absolute; opacity: 0;"
    tabindex="0"
  />
</body>
`;

exports[`Snapshot: CreateBlogPostWidget > Authorized & Mutation is loading 1`] = `
<body
  class="ReactModal__Body--open"
>
  <div
    data-react-modal-body-trap=""
    style="position: absolute; opacity: 0;"
    tabindex="0"
  />
  <div
    class="ReactModalPortal"
  />
  <div>
    <div
      aria-label="Create Blog Post Button"
      class="
                    fixed
                    bottom-24
                    left-1/2 
                    z-10
                    md:left-[95%] 
                    -translate-x-1/2
                    translate-y-1/2
                "
    >
      <div
        class="
                            absolute
                            w-64
                            min-h-0
                            h-12
                            -translate-y-full
                            -translate-x-1/2
                            left-1/2
                            md:-translate-x-full
                            md:left-full
                            overflow-hidden
                            -top-2
                        "
        style="height: 16rem;"
      >
        <div
          class="
                            top-full
                            absolute
                            w-full  
                            border-black
                            bg-gray-400
                            rounded-md
                            p-2
                            transition-all
                            after:absolute first-letter:
                            after:left-0
                            md:after:left-auto
                            after:right-0
                            after:my-0
                            after:mx-auto
                            after:w-0
                            after:h-0
                            after:border-t-[1rem]
                            after:border-solid
                            after:border-t-gray-400
                            after:border-l-[2rem]
                            after:border-r-[2rem]
                            after:border-x-transparent
                            after:transition-all
                        "
        >
          <p
            class="p-3 text-center md:text-left"
          >
            You do not have authorization to create posts
          </p>
        </div>
      </div>
      <button
        class="
                        bg-highlight-green hover:bg-gray-700 hover:scale-110 bg-highlight-green cursor-pointer transition-all 
                        aspect-square
                        w-16 
                        rounded-full 
                        after:block 
                        before:block 
                        after:bg-white 
                        before:bg-white
                        after:absolute 
                        before:absolute
                        after:top-1/2 
                        before:top-1/2 
                        after:left-1/2 
                        before:left-1/2 
                        after:-translate-x-1/2 
                        before:-translate-x-1/2 
                        after:-translate-y-1/2 
                        before:-translate-y-1/2 
                        after:h-8 
                        before:h-2 
                        after:w-2 
                        before:w-8 
                        after:transition-all 
                        before:transition-all 
                        "
      />
    </div>
  </div>
  <div
    class="ReactModalPortal"
  >
    <div
      class="ReactModal__Overlay fixed inset-0 z-20 bg-black/75"
    >
      <div
        aria-label="Blog Post Input"
        aria-modal="true"
        class="ReactModal__Content absolute top-1/2 left-1/2 z-40 w-11/12 -translate-x-1/2 -translate-y-1/2 rounded-md bg-white md:w-1/2 p-4"
        role="dialog"
        tabindex="-1"
      >
        <div
          class="flex pb-3 w-full flex-row items-center justify-between border-b border-b-gray-200 pl-4 pr-4"
        >
          <div
            class="font-semibold"
          >
            Create
             Blog Post
          </div>
        </div>
        <div
          class="flex h-96 w-full items-center justify-center"
        >
          <div
            class="inline-block animate-spin rounded-full border-solid border-current border-r-transparent align-[-0.125em] text-highlight-green motion-reduce:animate-[spin_1.5s_linear_infinite]"
            role="status"
            style="width: 12rem; height: 12rem; border-width: 1rem;"
          >
            <span
              class="absolute -m-px h-px w-px overflow-hidden whitespace-nowrap border-0 p-0 [clip:rect(0,0,0,0)]"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    data-react-modal-body-trap=""
    style="position: absolute; opacity: 0;"
    tabindex="0"
  />
</body>
`;

exports[`Snapshot: CreateBlogPostWidget > Authorized CreateBlogPostWidget 1`] = `
<body>
  <div>
    <div
      aria-label="Create Blog Post Button"
      class="
                    fixed
                    bottom-24
                    left-1/2 
                    z-10
                    md:left-[95%] 
                    -translate-x-1/2
                    translate-y-1/2
                "
    >
      <div
        class="
                            absolute
                            w-64
                            min-h-0
                            h-12
                            -translate-y-full
                            -translate-x-1/2
                            left-1/2
                            md:-translate-x-full
                            md:left-full
                            overflow-hidden
                            -top-2
                        "
        style="height: 16rem;"
      >
        <div
          class="
                            top-full
                            absolute
                            w-full  
                            border-black
                            bg-gray-400
                            rounded-md
                            p-2
                            transition-all
                            after:absolute first-letter:
                            after:left-0
                            md:after:left-auto
                            after:right-0
                            after:my-0
                            after:mx-auto
                            after:w-0
                            after:h-0
                            after:border-t-[1rem]
                            after:border-solid
                            after:border-t-gray-400
                            after:border-l-[2rem]
                            after:border-r-[2rem]
                            after:border-x-transparent
                            after:transition-all
                        "
        >
          <p
            class="p-3 text-center md:text-left"
          >
            You do not have authorization to create posts
          </p>
        </div>
      </div>
      <button
        class="
                        bg-highlight-green hover:bg-gray-700 hover:scale-110 bg-highlight-green cursor-pointer transition-all 
                        aspect-square
                        w-16 
                        rounded-full 
                        after:block 
                        before:block 
                        after:bg-white 
                        before:bg-white
                        after:absolute 
                        before:absolute
                        after:top-1/2 
                        before:top-1/2 
                        after:left-1/2 
                        before:left-1/2 
                        after:-translate-x-1/2 
                        before:-translate-x-1/2 
                        after:-translate-y-1/2 
                        before:-translate-y-1/2 
                        after:h-8 
                        before:h-2 
                        after:w-2 
                        before:w-8 
                        after:transition-all 
                        before:transition-all 
                        "
      />
    </div>
  </div>
  <div
    class="ReactModalPortal"
  />
</body>
`;

exports[`Snapshot: CreateBlogPostWidget > Loading CreateBlogPostWidget 1`] = `
<body>
  <div>
    <div
      aria-label="Create Blog Post Button"
      class="
                    fixed
                    bottom-24
                    left-1/2 
                    z-10
                    md:left-[95%] 
                    -translate-x-1/2
                    translate-y-1/2
                "
    >
      <div
        class="inline-block animate-spin rounded-full border-solid border-current border-r-transparent align-[-0.125em] text-highlight-green motion-reduce:animate-[spin_1.5s_linear_infinite]"
        role="status"
        style="width: 4rem; height: 4rem; border-width: 0.3333333333333333rem;"
      >
        <span
          class="absolute -m-px h-px w-px overflow-hidden whitespace-nowrap border-0 p-0 [clip:rect(0,0,0,0)]"
        />
      </div>
    </div>
  </div>
</body>
`;

exports[`Snapshot: CreateBlogPostWidget > Unauthorized Plain CreateBlogPostWidget 1`] = `
<body>
  <div>
    <div
      aria-label="Create Blog Post Button"
      class="
                    fixed
                    bottom-24
                    left-1/2 
                    z-10
                    md:left-[95%] 
                    -translate-x-1/2
                    translate-y-1/2
                "
    >
      <div
        class="
                            absolute
                            w-64
                            min-h-0
                            h-12
                            -translate-y-full
                            -translate-x-1/2
                            left-1/2
                            md:-translate-x-full
                            md:left-full
                            overflow-hidden
                            -top-2
                        "
        style="height: 16rem;"
      >
        <div
          class="
                            top-full
                            absolute
                            w-full  
                            border-black
                            bg-gray-400
                            rounded-md
                            p-2
                            transition-all
                            after:absolute first-letter:
                            after:left-0
                            md:after:left-auto
                            after:right-0
                            after:my-0
                            after:mx-auto
                            after:w-0
                            after:h-0
                            after:border-t-[1rem]
                            after:border-solid
                            after:border-t-gray-400
                            after:border-l-[2rem]
                            after:border-r-[2rem]
                            after:border-x-transparent
                            after:transition-all
                        "
        >
          <p
            class="p-3 text-center md:text-left"
          >
            You do not have authorization to create posts
          </p>
        </div>
      </div>
      <button
        class="
                        bg-gray-200 cursor-not-allowed 
                        aspect-square
                        w-16 
                        rounded-full 
                        after:block 
                        before:block 
                        after:bg-white 
                        before:bg-white
                        after:absolute 
                        before:absolute
                        after:top-1/2 
                        before:top-1/2 
                        after:left-1/2 
                        before:left-1/2 
                        after:-translate-x-1/2 
                        before:-translate-x-1/2 
                        after:-translate-y-1/2 
                        before:-translate-y-1/2 
                        after:h-8 
                        before:h-2 
                        after:w-2 
                        before:w-8 
                        after:transition-all 
                        before:transition-all 
                        "
      />
    </div>
  </div>
</body>
`;
